fix AsApproximateFloat64() for BinarySI

y-tag opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug


#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

Fixes #103748

#### Special notes for your reviewer:

Since `Format`, such as `DecimalSI` and `BinarySI`, would be used for conversion from/to string, the value of `AsApproximateFloat64()` does not depend on `Format`.

After applying this patch, the result of `main.go` in #103748 changes as follow: 
```
$ go run main.go 
s: 3750, v: 3750, f: 3750
s: 7500k, v: 7500000, f: 7.5e+06
s: 2250M, v: 2250000000, f: 2.25e+09
s: 3840, v: 3840, f: 3840.0000000000005
s: 7680Ki, v: 7864320, f: 7.864320000000001e+06
s: 2304Mi, v: 2415919104, f: 2.415919104e+09
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Scheduler resource metrics over fractional binary quantities (2.5Gi, 1.1Ki) were incorrectly reported as very small values.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

y-tag committed:
fix AsApproximateFloat64() for BinarySI

fedebongio commented:
/assign @yliaog 
/triage accepted

yliaog commented:
/ok-to-test

y-tag commented:
/retest

yliaog commented:
/lgtm

y-tag commented:
/assign @deads2k

y-tag commented:
/unassign @deads2k
/assign @liggitt

y-tag commented:
@liggitt I would be grateful for your approval of this PR.

liggitt commented:
/unassign
/assign @smarterclayton 
as original author (https://github.com/kubernetes/kubernetes/commit/c06c1121d1c07f72cf0d35c76a1637814efe9144)

smarterclayton committed:
Additional resource quantity testing

Fractional binary SI quantities that cannot be represented as decimal
internally were incorrectly calculated.

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go:
```diff
@@ -1207,11 +1207,11 @@ func TestQuantityAsApproximateFloat64(t *testing.T) {
+		{decQuantity(7*1024*1024, 4, BinarySI), (7 * 1024 * 1024) * 10000},
+		{decQuantity(7*1024*1024, 8, BinarySI), (7 * 1024 * 1024) * 100000000},
+		{decQuantity(7*1024*1024, -1, BinarySI), (7 * 1024 * 1024) * math.Pow10(-1)}, // '* Pow10' and '/ float(10)' do not round the same way
+		{decQuantity(7*1024*1024, -8, BinarySI), (7 * 1024 * 1024) / float64(100000000)},
```
Looks correct, please pull in the changes in https://github.com/kubernetes/kubernetes/pull/104757 (the additional test added to quantity_test.go and the change to scheduler resources_test.go to surface the problem).

The release note needs to be changed to read:

> Scheduler resource metrics over fractional binary quantities (`2.5Gi`, `1.1Ki`) were incorrectly reported using the wrong power.

y-tag replied:
I appreciate your help! 🙇
I pulled in the changes in #104757 and updated the release note.

smarterclayton commented:
Looks good except for the additional test changes.  Once those are integrated and the release note updated I'll tag.

ehashman commented:
/priority important-soon

y-tag commented:
@smarterclayton Could you check if I have followed your instructions?
I added your commit to this PR and updated the release note. Thank you!

smarterclayton commented:
/lgtm
/approve

Thanks!
